<html>
<head><meta charset="utf-8" />
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/codemirror/5.65.3/codemirror.min.css" crossorigin="anonymous" referrerpolicy="no-referrer" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.15.6/katex.min.css" crossorigin="anonymous" referrerpolicy="no-referrer" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/antd/4.22.8/antd.min.css" crossorigin="anonymous" referrerpolicy="no-referrer" />
</head>
<body>
<div style="margin:30px 30px 0 0"><div style="margin-bottom:15px"><b>Kernel:</b> <!-- -->R</div><div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><h1 id="coo---single-cell-rna-sequencing-data-analysis"><span>COO - single cell RNA sequencing data analysis</span></h1><h2 id="bioinformatics-in-neuroscience-course-2022"><span>Bioinformatics in Neuroscience course, 2022</span></h2></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><p><span style="text-indent:0"><span></span><img src="https://cocalc7.science.uu.nl/projects/510bc61c-e9f3-40bc-8f23-1a94536e021a/files/scRNAseq_human_brain_2022/extra/COO_DALLE_2022-09-27_10.21.35_small.png" alt="Single cell data analysis of the adult human brain" title="" style="max-width:100%;max-height:100%;object-fit:cover"/><span></span></span></p></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><div><div><div class="alert alert-warning"><b>About this tutorial</b><br/><br/>
<p>
This is the single cell RNA sequencing (scRNAseq) data analysis tutorial, prepared for the scRNAseq days of the Bioinformatics in Neuroscience course, organised for the Neuroscience and Cognition master  and other Utrecht University master programs. The script follows the well known vignette of the Seurat (v3) pipeline for pbmc, adapted to analyse the scRNAseq data of the adult human brain.
</p><p>
</p></div> 
<div class="">
    <b>The data </b> is a subset of the recent work from Siletti et al (Bioarxiv, 2022; https://www.biorxiv.org/content/10.1101/2022.10.12.511898v1), which you received for this course. In this study, ~4M nuclei from different parts of the human brain is sequenced. You will later here a talk by the first author of the study, which will be a natural continuation of this practice. 
A subset of the original atlas is provided for the purpose of this practical in the file <b><i>&quot;../rawdata/A46_10X176_8_count_data.rds&quot; </i></b>. 
An accompanying metadata file,<b><i> &quot;../rawdata/A46_10X176_8_metadata.rds&quot;</i></b>, will give you additional information. The rawdata as well as analysis of the entire dataset can be reached at: https://github.com/linnarsson-lab/adult-human-brain 
<br/><br/>
<b>Participants will then learn </b> how to use the ‘Seurat’ pipeline, the most popular ‘R’ package to process and analyze scRNAseq data. The instructor will guide the students step-by-step, where a prewritten code will take through the first steps of scRNAseq data analysis. These include shaping the raw data, preprocessing, dimensionality reduction, basic clustering and differential gene expression analysis. 
    <br/><br/>
</div></div></div><div><div><div class="alert alert-success"><b>Learning goals</b><br/><br/>
<p>1. Gain the capacity to run a standard scRNAseq data analysis pipeline independently</p>
<p>2. Have the ability to interpret the results of single cell data analysis</p>
<p>
</p></div></div></div><p><span style="text-indent:0"><span><strong>The course</strong></span><span> used the CoCalc server: </span><a href="https://cocalc7.science.uu.nl/" target="_blank" rel="noopener"><span>https://cocalc7.science.uu.nl/</span></a><span> </span></span></p></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><div><div><div class="alert alert-success"><b>During the tutorial, keep in mind</b><br/><br/>
<p>1. What are the major considerations during the preprocessing steps?</p>
<p>2. How does dimensionality reduction help us understand the data?</p>
<p>3. What is a molecular signature? How can marker genes help us undertand cellular (and brain) function?</p>
</div></div></div></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><div><div><div class="alert alert-warning"><h2>Part I: Upload the data and create a Seurat object</h2></div></div></div></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><p><span style="text-indent:0"><span><strong>Some good practice</strong></span><span style="display:inline"><span><br/></span></span><span><span style="white-space:normal"> </span><span></span></span><span>To keep everything ordered, make some folders for saving the data and the figures. You will receive a warning if you have already made these folders</span></span></p></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px">In [<!-- -->0<!-- -->]:</div><div style="width:100%;border-radius:2px;position:relative;overflow-x:auto;border:1px solid rgb(207, 207, 207)"><pre class="CodeMirror cm-s-default  CodeMirror-wrap" style="width:100%;overflow-x:auto;line-height:normal;height:auto;font-size:inherit;margin-bottom:0;padding:4px;white-space:pre-wrap;word-wrap:break-word;word-break:normal;border:0;background:white"><div style="margin-left:0"><span class="cm-variable">dir.create</span><span>(</span><span class="cm-string">&quot;</span><span class="cm-string">write/&quot;</span><span>)</span><span>
</span><span class="cm-variable">dir.create</span><span>(</span><span class="cm-string">&quot;</span><span class="cm-string">figures/&quot;</span><span>)</span><span>
</span></div></pre></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><h3 id="11-load-libraries"><span>1.1 Load libraries</span></h3><p><span style="text-indent:0"><span>Load the necessary libraries to run our pipeline based on the well-known Seurat pipeline. You can find detailed information on Seurat as well as tutorials on its website (</span><a href="https://satijalab.org/seurat/" target="_blank" rel="noopener"><span>https://satijalab.org/seurat/</span></a><span>) or on Github (</span><a href="https://github.com/satijalab/seurat" target="_blank" rel="noopener"><span>https://github.com/satijalab/seurat</span></a><span>)</span></span></p><p><span style="text-indent:0"><span>We are going to use Seurat v3, which is well established. Note that there is a recent beta release (v4) with additional functionalities</span></span></p></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px">In [<!-- -->0<!-- -->]:</div><div style="width:100%;border-radius:2px;position:relative;overflow-x:auto;border:1px solid rgb(207, 207, 207)"><pre class="CodeMirror cm-s-default  CodeMirror-wrap" style="width:100%;overflow-x:auto;line-height:normal;height:auto;font-size:inherit;margin-bottom:0;padding:4px;white-space:pre-wrap;word-wrap:break-word;word-break:normal;border:0;background:white"><div style="margin-left:0"><span class="cm-comment"># Load the libraries</span><span>
</span><span class="cm-variable">library</span><span>(</span><span class="cm-variable">Seurat</span><span>,</span><span class="cm-variable">verbose</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-atom">FALSE</span><span>)</span><span>
</span><span class="cm-variable">library</span><span>(</span><span class="cm-variable">ggplot2</span><span>,</span><span class="cm-variable">verbose</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-atom">FALSE</span><span>)</span><span>
</span><span class="cm-variable">library</span><span>(</span><span class="cm-variable">RColorBrewer</span><span>,</span><span class="cm-variable">verbose</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-atom">FALSE</span><span>)</span><span>
</span><span class="cm-variable">library</span><span>(</span><span class="cm-variable">useful</span><span>,</span><span class="cm-variable">verbose</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-atom">FALSE</span><span>)</span><span>
</span><span class="cm-variable">library</span><span>(</span><span class="cm-variable">readr</span><span>,</span><span class="cm-variable">verbose</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-atom">FALSE</span><span>)</span><span>
</span><span class="cm-variable">library</span><span>(</span><span class="cm-variable">hdf5r</span><span>,</span><span class="cm-variable">verbose</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-atom">FALSE</span><span>)</span><span>
</span><span>
</span><span class="cm-comment"># check if the following are on</span><span>
</span><span class="cm-variable">library</span><span>(</span><span class="cm-variable">dplyr</span><span>,</span><span class="cm-variable">verbose</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-atom">FALSE</span><span>)</span><span>
</span><span class="cm-variable">library</span><span>(</span><span class="cm-variable">stringr</span><span>,</span><span class="cm-variable">verbose</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-atom">FALSE</span><span>)</span><span>
</span><span>
</span></div></pre></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><h3 id="12-load-the-data"><span>1.2 Load the data</span></h3><p><span style="text-indent:0"><span>We will use the 10x Genomics data generated from the adult human brain. Since 4M cells are too many, we are using a ~5k subset. The data is stored in two seperate files</span></span></p></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px">In [<!-- -->0<!-- -->]:</div><div style="width:100%;border-radius:2px;position:relative;overflow-x:auto;border:1px solid rgb(207, 207, 207)"><pre class="CodeMirror cm-s-default  CodeMirror-wrap" style="width:100%;overflow-x:auto;line-height:normal;height:auto;font-size:inherit;margin-bottom:0;padding:4px;white-space:pre-wrap;word-wrap:break-word;word-break:normal;border:0;background:white"><div style="margin-left:0"><span class="cm-comment"># This contains the actual data as a cell x gene matrix. Each cell in this matrix contain information on how many reads we have for a given gene in a given cell</span><span>
</span><span class="cm-variable">count_data</span><span> </span><span class="cm-operator arrow">&lt;-</span><span> </span><span class="cm-variable">read_rds</span><span>(</span><span class="cm-string">&quot;</span><span class="cm-string">../rawdata/A46_10X176_8_count_data.rds&quot;</span><span>)</span><span>
</span></div></pre></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px">In [<!-- -->0<!-- -->]:</div><div style="width:100%;border-radius:2px;position:relative;overflow-x:auto;border:1px solid rgb(207, 207, 207)"><pre class="CodeMirror cm-s-default  CodeMirror-wrap" style="width:100%;overflow-x:auto;line-height:normal;height:auto;font-size:inherit;margin-bottom:0;padding:4px;white-space:pre-wrap;word-wrap:break-word;word-break:normal;border:0;background:white"><div style="margin-left:0"><span class="cm-comment"># the dimensions for a sanity check. The first dimension shows the number of features (in this case, genes) while the second indicates teh number of cells</span><span>
</span><span class="cm-variable">dim</span><span>(</span><span class="cm-variable">count_data</span><span>)</span><span>
</span></div></pre></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px">In [<!-- -->0<!-- -->]:</div><div style="width:100%;border-radius:2px;position:relative;overflow-x:auto;border:1px solid rgb(207, 207, 207)"><pre class="CodeMirror cm-s-default  CodeMirror-wrap" style="width:100%;overflow-x:auto;line-height:normal;height:auto;font-size:inherit;margin-bottom:0;padding:4px;white-space:pre-wrap;word-wrap:break-word;word-break:normal;border:0;background:white"><div style="margin-left:0"><span class="cm-comment"># This contains the metadata. Each column contains information for each cell in teh dataset.</span><span>
</span><span class="cm-variable">metadata</span><span> </span><span class="cm-operator arrow">&lt;-</span><span>  </span><span class="cm-variable">read_rds</span><span>(</span><span class="cm-string">&#x27;</span><span class="cm-string">../rawdata/A46_10X176_8_metadata.rds&#x27;</span><span>)</span><span>
</span></div></pre></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px">In [<!-- -->0<!-- -->]:</div><div style="width:100%;border-radius:2px;position:relative;overflow-x:auto;border:1px solid rgb(207, 207, 207)"><pre class="CodeMirror cm-s-default  CodeMirror-wrap" style="width:100%;overflow-x:auto;line-height:normal;height:auto;font-size:inherit;margin-bottom:0;padding:4px;white-space:pre-wrap;word-wrap:break-word;word-break:normal;border:0;background:white"><div style="margin-left:0"><span class="cm-comment"># what do we have in the metadata?</span><span>
</span><span class="cm-variable">colnames</span><span>(</span><span class="cm-variable">metadata</span><span>)</span><span>
</span></div></pre></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><h3 id="13-create-seurat-object"><span>1.3 Create Seurat object</span></h3></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><p><span style="text-indent:0"><span>Here, we will create a &quot;seurat object&quot; which has a different structure than a data table. Seurat object contains the data as a matrix in a separate slot, but can also have &#x27;features&#x27; that are linked to genes or cells. For instance cell types for each cell can be stored in a separate slot. In addition, we can store additional information e.g. results of differential expression test or coordinates for each cell after a given dimensionality reduction (e.g. pca)</span></span></p></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><p><span style="text-indent:0"><span>&#x27;min.cells&#x27; is to filter </span><span><em>genes</em></span><span>: only genes detected in min of these number of cells will be kept</span><span><span style="white-space:normal"> </span><span></span></span><span>&#x27;min.features&#x27; is to filter </span><span><em>cells</em></span><span>: only cells that have a minimum total of given number of features will be kept</span><span><span style="white-space:normal"> </span><span></span></span><span>Thus, these two provide initial quality control of the data</span></span></p></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px">In [<!-- -->0<!-- -->]:</div><div style="width:100%;border-radius:2px;position:relative;overflow-x:auto;border:1px solid rgb(207, 207, 207)"><pre class="CodeMirror cm-s-default  CodeMirror-wrap" style="width:100%;overflow-x:auto;line-height:normal;height:auto;font-size:inherit;margin-bottom:0;padding:4px;white-space:pre-wrap;word-wrap:break-word;word-break:normal;border:0;background:white"><div style="margin-left:0"><span class="cm-variable">dataset</span><span> </span><span class="cm-operator arrow">&lt;-</span><span> </span><span class="cm-variable">CreateSeuratObject</span><span>(</span><span class="cm-variable">counts</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">count_data</span><span>,</span><span> </span><span class="cm-variable">min.cells</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-number">1</span><span>,</span><span> </span><span class="cm-variable">min.features</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-number">1000</span><span>,</span><span> </span><span class="cm-variable">meta.data</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">metadata</span><span>,</span><span>
</span><span>                          </span><span class="cm-variable">project</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-string">&quot;</span><span class="cm-string">Human_brain&quot;</span><span>)</span><span>
</span></div></pre></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><p><span style="text-indent:0"><span>visualise the number of genes and features, which are already provided in the dataset</span></span></p></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px">In [<!-- -->0<!-- -->]:</div><div style="width:100%;border-radius:2px;position:relative;overflow-x:auto;border:1px solid rgb(207, 207, 207)"><pre class="CodeMirror cm-s-default  CodeMirror-wrap" style="width:100%;overflow-x:auto;line-height:normal;height:auto;font-size:inherit;margin-bottom:0;padding:4px;white-space:pre-wrap;word-wrap:break-word;word-break:normal;border:0;background:white"><div style="margin-left:0"><span class="cm-variable">head</span><span>(</span><span class="cm-variable">dataset</span><span>@</span><span class="cm-variable">meta.data</span><span>)</span><span>
</span></div></pre></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><p><span style="text-indent:0"><span>The first visual quality control is to look at the number of cells (nCount_RNA) and number of genes (nFeature_RNA) per cells. These features are calculated while generating the Seurat object. For instance, the number of RNA molecules per cell is stored as &#x27;nCount_RNA&#x27;, which is located in dataset@meta.data$nCount_RNA</span></span></p></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><h3 id="14-visualise-basic-metrics"><span>1.4 Visualise basic metrics</span></h3></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px">In [<!-- -->0<!-- -->]:</div><div style="width:100%;border-radius:2px;position:relative;overflow-x:auto;border:1px solid rgb(207, 207, 207)"><pre class="CodeMirror cm-s-default  CodeMirror-wrap" style="width:100%;overflow-x:auto;line-height:normal;height:auto;font-size:inherit;margin-bottom:0;padding:4px;white-space:pre-wrap;word-wrap:break-word;word-break:normal;border:0;background:white"><div style="margin-left:0"><span class="cm-variable">plot1</span><span> </span><span class="cm-operator arrow">&lt;-</span><span> </span><span class="cm-variable">VlnPlot</span><span>(</span><span class="cm-variable">object</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">dataset</span><span>,</span><span> </span><span class="cm-variable">features</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">c</span><span>(</span><span class="cm-string">&quot;</span><span class="cm-string">nCount_RNA&quot;</span><span>,</span><span> </span><span class="cm-string">&quot;</span><span class="cm-string">nFeature_RNA&quot;</span><span>)</span><span>,</span><span class="cm-variable">pt.size</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-number">.01</span><span>)</span><span>
</span><span class="cm-variable">plot2</span><span> </span><span class="cm-operator arrow">&lt;-</span><span> </span><span class="cm-variable">FeatureScatter</span><span>(</span><span class="cm-variable">dataset</span><span>,</span><span> </span><span class="cm-variable">feature1</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-string">&quot;</span><span class="cm-string">nCount_RNA&quot;</span><span>,</span><span> </span><span class="cm-variable">feature2</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-string">&quot;</span><span class="cm-string">nFeature_RNA&quot;</span><span>)</span><span>
</span><span class="cm-variable">plot1</span><span> </span><span class="cm-operator">|</span><span> </span><span class="cm-variable">plot2</span><span>
</span></div></pre></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><div><div><div class="alert alert-warning"><b>Question 1.1
<p>
What do you see? :</p></b> 
<p>1. How do the number of genes and RNA molecules correlate?  </p>
<p>2. There are two populations visible on the violin plot... What do these mean? </p>
<p><b>Answer whenever you want!</b></p>
</div></div></div></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><div><div><div class="alert alert-danger"><h2>Back to the presentation</h2></div></div></div></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><div><div><div class="alert alert-success"><h2>Part II: Preprocessing</h2>
Quality control</div></div></div></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><h2 id="21-standard-pre-processing-workflow"><span>2.1 Standard pre-processing workflow</span></h2></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><p><span style="text-indent:0"><span>set a name that will be used later for the files you save. Anything goes. You can add your own name</span></span></p></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px">In [<!-- -->0<!-- -->]:</div><div style="width:100%;border-radius:2px;position:relative;overflow-x:auto;border:1px solid rgb(207, 207, 207)"><pre class="CodeMirror cm-s-default  CodeMirror-wrap" style="width:100%;overflow-x:auto;line-height:normal;height:auto;font-size:inherit;margin-bottom:0;padding:4px;white-space:pre-wrap;word-wrap:break-word;word-break:normal;border:0;background:white"><div style="margin-left:0"><span class="cm-variable">name</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-string">&quot;</span><span class="cm-string">my_analysis_name&quot;</span><span>
</span></div></pre></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><h3 id="211-calculate-additional-qc-metrics"><span>2.1.1 Calculate additional QC metrics</span></h3></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><div><div><h5>
    General information on the use of mitochondria and ribosomal proteins for QC
</h5>
Mitochondrial RNA is the mRNA that is generated by the mitochondrial genome. Normally, these constitute a small fraction of the total mRNA. However, in dying or damaged cells, while cytoplasmic/nuclear mRNA degrades rapidly, mitochondrial mRNA is rather well preserved. Thus, a high ratio of mitochondrial mRNA indicates BAD cell quality.</div></div><p><span style="text-indent:0"><span>mRNA coding for the Ribosomal subunit proteins is abundant (not to be confused with rRNA). Usually, a high ribosomal RNA percentage indicates production of a lot of proteins, and is very high in dividing cells or some secretory cells that need to constantly produce proteins. However, if most of the mRNA (&gt;30-50%) that we detect is ribosomal, it means that we wont learn much from this cell, and that we should exclude it from the analysis.</span></span></p></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><div><div><div class="alert alert-warning"><b>Question 2.1
<p>
Note that we are analysing single nuclei data. We have the following question:</p></b> 
<p>1. How do you think this will effect our quality control? </p>
<p>2. In this case, what does high ribosomal/mitochondrial percentage mean?</p>
<p>Spend  1 min to think about the questions. Then, talk to the person next to you about what you think. Finally, summarise it in one sentence</p>
</div></div></div></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px">In [<!-- -->0<!-- -->]:</div><div style="width:100%;border-radius:2px;position:relative;overflow-x:auto;border:1px solid rgb(207, 207, 207)"><pre class="CodeMirror cm-s-default  CodeMirror-wrap" style="width:100%;overflow-x:auto;line-height:normal;height:auto;font-size:inherit;margin-bottom:0;padding:4px;white-space:pre-wrap;word-wrap:break-word;word-break:normal;border:0;background:white"><div style="margin-left:0"><span class="cm-comment"># Start by generating  QC metrics additional to the no of genes/features</span><span>
</span><span class="cm-variable">dataset</span><span> </span><span class="cm-operator arrow">&lt;-</span><span> </span><span class="cm-variable">PercentageFeatureSet</span><span>(</span><span class="cm-variable">dataset</span><span>,</span><span class="cm-variable">pattern</span><span class="cm-operator">=</span><span class="cm-string">&#x27;</span><span class="cm-string">^MT-&#x27;</span><span>,</span><span> </span><span class="cm-variable">col.name</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-string">&quot;</span><span class="cm-string">percent.mt&quot;</span><span>)</span><span> </span><span class="cm-comment">#Select all features starting with MT-, which are the Mitochondrial genes</span><span>
</span><span class="cm-variable">dataset</span><span> </span><span class="cm-operator arrow">&lt;-</span><span> </span><span class="cm-variable">PercentageFeatureSet</span><span>(</span><span class="cm-variable">dataset</span><span>,</span><span class="cm-variable">pattern</span><span class="cm-operator">=</span><span class="cm-string">&#x27;</span><span class="cm-string">RP(S|L)&#x27;</span><span>,</span><span> </span><span class="cm-variable">col.name</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-string">&quot;</span><span class="cm-string">percent.ribo&quot;</span><span>)</span><span> </span><span class="cm-comment"># Select all that starts with RPS or RPL</span><span>
</span></div></pre></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><p><span style="text-indent:0"><span>Plot and inspect. You can remove the ots in case they cause too much moise</span></span></p></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><h3 id="212-visualise-qc-metrics"><span>2.1.2 Visualise QC metrics</span></h3></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px">In [<!-- -->0<!-- -->]:</div><div style="width:100%;border-radius:2px;position:relative;overflow-x:auto;border:1px solid rgb(207, 207, 207)"><pre class="CodeMirror cm-s-default  CodeMirror-wrap" style="width:100%;overflow-x:auto;line-height:normal;height:auto;font-size:inherit;margin-bottom:0;padding:4px;white-space:pre-wrap;word-wrap:break-word;word-break:normal;border:0;background:white"><div style="margin-left:0"><span class="cm-variable">options</span><span>(</span><span class="cm-variable">repr.plot.width</span><span class="cm-operator">=</span><span class="cm-number">12</span><span>,</span><span> </span><span class="cm-variable">repr.plot.height</span><span class="cm-operator">=</span><span class="cm-number">6</span><span>)</span><span>
</span><span class="cm-variable">VlnPlot</span><span>(</span><span class="cm-variable">object</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">dataset</span><span>,</span><span> </span><span class="cm-variable">features</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">c</span><span>(</span><span class="cm-string">&quot;</span><span class="cm-string">percent.mt&quot;</span><span>,</span><span> </span><span class="cm-string">&quot;</span><span class="cm-string">percent.ribo&quot;</span><span>)</span><span>)</span><span>
</span></div></pre></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><p><span style="text-indent:0"><span>You can save them all as a pdf file to use in presentations, reports etc</span></span></p></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px">In [<!-- -->0<!-- -->]:</div><div style="width:100%;border-radius:2px;position:relative;overflow-x:auto;border:1px solid rgb(207, 207, 207)"><pre class="CodeMirror cm-s-default  CodeMirror-wrap" style="width:100%;overflow-x:auto;line-height:normal;height:auto;font-size:inherit;margin-bottom:0;padding:4px;white-space:pre-wrap;word-wrap:break-word;word-break:normal;border:0;background:white"><div style="margin-left:0"><span class="cm-variable">pdf</span><span>(</span><span class="cm-variable">paste0</span><span>(</span><span class="cm-variable">name</span><span>,</span><span class="cm-string">&quot;</span><span class="cm-string">_Vlnplot_stats_all.pdf&quot;</span><span>)</span><span>,</span><span class="cm-variable">height</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-number">6</span><span>,</span><span class="cm-variable">width</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-number">12</span><span>)</span><span>
</span><span class="cm-variable">VlnPlot</span><span>(</span><span class="cm-variable">object</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">dataset</span><span>,</span><span> </span><span class="cm-variable">features</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">c</span><span>(</span><span class="cm-string">&quot;</span><span class="cm-string">nCount_RNA&quot;</span><span>,</span><span> </span><span class="cm-string">&quot;</span><span class="cm-string">nFeature_RNA&quot;</span><span>,</span><span class="cm-string">&quot;</span><span class="cm-string">percent.mt&quot;</span><span>,</span><span> </span><span class="cm-string">&quot;</span><span class="cm-string">percent.ribo&quot;</span><span>)</span><span>)</span><span>
</span><span class="cm-comment"># If there are too many points for your liking, you can turn them off as follows:</span><span>
</span><span class="cm-comment">#VlnPlot(object = dataset, features = c(&quot;nCount_RNA&quot;, &quot;nFeature_RNA&quot;,&quot;percent.mt&quot;, &quot;percent.ribo&quot;),pt.size = 0, cols = &quot;blue&quot;)</span><span>
</span><span class="cm-variable">dev.off</span><span>(</span><span>)</span><span>
</span><span>
</span></div></pre></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><p><span style="text-indent:0"><span>Visualise multiple metrics for each cell as a measure of the number of reads (in other words; UMI, or unique RNA molecules). This helps use to evaluate the quality of the data in multiple ways</span></span></p></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px">In [<!-- -->0<!-- -->]:</div><div style="width:100%;border-radius:2px;position:relative;overflow-x:auto;border:1px solid rgb(207, 207, 207)"><pre class="CodeMirror cm-s-default  CodeMirror-wrap" style="width:100%;overflow-x:auto;line-height:normal;height:auto;font-size:inherit;margin-bottom:0;padding:4px;white-space:pre-wrap;word-wrap:break-word;word-break:normal;border:0;background:white"><div style="margin-left:0"><span class="cm-variable">plot1</span><span> </span><span class="cm-operator arrow">&lt;-</span><span> </span><span class="cm-variable">FeatureScatter</span><span>(</span><span class="cm-variable">dataset</span><span>,</span><span> </span><span class="cm-variable">feature1</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-string">&quot;</span><span class="cm-string">nCount_RNA&quot;</span><span>,</span><span> </span><span class="cm-variable">feature2</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-string">&quot;</span><span class="cm-string">percent.mt&quot;</span><span>)</span><span>
</span><span class="cm-variable">plot2</span><span> </span><span class="cm-operator arrow">&lt;-</span><span> </span><span class="cm-variable">FeatureScatter</span><span>(</span><span class="cm-variable">dataset</span><span>,</span><span> </span><span class="cm-variable">feature1</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-string">&quot;</span><span class="cm-string">nCount_RNA&quot;</span><span>,</span><span> </span><span class="cm-variable">feature2</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-string">&quot;</span><span class="cm-string">percent.ribo&quot;</span><span>)</span><span>
</span><span class="cm-variable">plot1</span><span> </span><span class="cm-operator">|</span><span> </span><span class="cm-variable">plot2</span><span>
</span><span>
</span></div></pre></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><h3 id="22-filter"><span>2.2 Filter</span></h3></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><p><span style="text-indent:0"><span>Decide on the filtering parameters that you want to use. This is not always objective! However, you can see the &#x27;outliers&#x27; in the data rather well</span></span></p></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px">In [<!-- -->0<!-- -->]:</div><div style="width:100%;border-radius:2px;position:relative;overflow-x:auto;border:1px solid rgb(207, 207, 207)"><pre class="CodeMirror cm-s-default  CodeMirror-wrap" style="width:100%;overflow-x:auto;line-height:normal;height:auto;font-size:inherit;margin-bottom:0;padding:4px;white-space:pre-wrap;word-wrap:break-word;word-break:normal;border:0;background:white"><div style="margin-left:0"><span class="cm-comment">#Now filter</span><span>
</span><span class="cm-variable">dataset</span><span> </span><span class="cm-operator arrow">&lt;-</span><span> </span><span class="cm-variable">subset</span><span>(</span><span class="cm-variable">x</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">dataset</span><span>,</span><span> </span><span class="cm-variable">subset</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">percent.mt</span><span> </span><span class="cm-operator">&lt;</span><span> </span><span class="cm-number">5</span><span> </span><span class="cm-operator">&amp;</span><span> </span><span class="cm-variable">nCount_RNA</span><span> </span><span class="cm-operator">&lt;</span><span> </span><span class="cm-number">90000</span><span> </span><span class="cm-operator">&amp;</span><span> </span><span class="cm-variable">percent.ribo</span><span> </span><span class="cm-operator">&lt;</span><span> </span><span class="cm-number">2</span><span>)</span><span>
</span></div></pre></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px">In [<!-- -->0<!-- -->]:</div><div style="width:100%;border-radius:2px;position:relative;overflow-x:auto;border:1px solid rgb(207, 207, 207)"><pre class="CodeMirror cm-s-default  CodeMirror-wrap" style="width:100%;overflow-x:auto;line-height:normal;height:auto;font-size:inherit;margin-bottom:0;padding:4px;white-space:pre-wrap;word-wrap:break-word;word-break:normal;border:0;background:white"><div style="margin-left:0"><span class="cm-variable">dataset</span><span>
</span></div></pre></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><p><span style="text-indent:0"><span>Now get rid of the mitochondrial genes which we do not need for downstream analysis.</span></span></p></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px">In [<!-- -->0<!-- -->]:</div><div style="width:100%;border-radius:2px;position:relative;overflow-x:auto;border:1px solid rgb(207, 207, 207)"><pre class="CodeMirror cm-s-default  CodeMirror-wrap" style="width:100%;overflow-x:auto;line-height:normal;height:auto;font-size:inherit;margin-bottom:0;padding:4px;white-space:pre-wrap;word-wrap:break-word;word-break:normal;border:0;background:white"><div style="margin-left:0"><span class="cm-variable">dataset</span><span> </span><span class="cm-operator arrow">&lt;-</span><span> </span><span class="cm-variable">subset</span><span>(</span><span class="cm-variable">x</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">dataset</span><span>[</span><span class="cm-operator">-</span><span class="cm-variable">grep</span><span>(</span><span class="cm-variable">pattern</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-string">&quot;</span><span class="cm-string">^MT-&quot;</span><span>,</span><span> </span><span class="cm-variable">x</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">rownames</span><span>(</span><span class="cm-variable">dataset</span><span>)</span><span>,</span><span> </span><span class="cm-variable">value</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-atom">FALSE</span><span>)</span><span>]</span><span>)</span><span>
</span><span class="cm-variable">dataset</span><span> </span><span>
</span></div></pre></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px">In [<!-- -->0<!-- -->]:</div><div style="width:100%;border-radius:2px;position:relative;overflow-x:auto;border:1px solid rgb(207, 207, 207)"><pre class="CodeMirror cm-s-default  CodeMirror-wrap" style="width:100%;overflow-x:auto;line-height:normal;height:auto;font-size:inherit;margin-bottom:0;padding:4px;white-space:pre-wrap;word-wrap:break-word;word-break:normal;border:0;background:white"><div style="margin-left:0"><span>
</span></div></pre></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><div><div><div class="alert alert-danger"><h2>Back to the presentation</h2></div></div></div></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><div><div><div class="alert alert-success"><h2>Part III: Normalize and find variable features</h2>
Quality control</div></div></div></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><h3 id="31-normalisation"><span>3.1 Normalisation</span></h3></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><p><span style="text-indent:0"><span>Standard flow of Seurat is replaced by a single command. However, it is good to see this part to learn each step</span></span></p></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px">In [<!-- -->0<!-- -->]:</div><div style="width:100%;border-radius:2px;position:relative;overflow-x:auto;border:1px solid rgb(207, 207, 207)"><pre class="CodeMirror cm-s-default  CodeMirror-wrap" style="width:100%;overflow-x:auto;line-height:normal;height:auto;font-size:inherit;margin-bottom:0;padding:4px;white-space:pre-wrap;word-wrap:break-word;word-break:normal;border:0;background:white"><div style="margin-left:0"><span class="cm-comment">## normalise the data</span><span>
</span><span class="cm-variable">dataset</span><span> </span><span class="cm-operator arrow">&lt;-</span><span> </span><span class="cm-variable">NormalizeData</span><span>(</span><span class="cm-variable">object</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">dataset</span><span>,</span><span> </span><span class="cm-variable">normalization.method</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-string">&quot;</span><span class="cm-string">LogNormalize&quot;</span><span>,</span><span> </span><span class="cm-variable">scale.factor</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-number">10000</span><span>)</span><span>
</span></div></pre></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><p><span style="text-indent:0"><span>The log transformed RNA counts from each cell is now normalised using a size factor. In other words, the sum of counts for each cells is more or less the same. We can now compare different cells to each other</span></span></p></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><h3 id="32-detection-of-variable-genes-across-the-single-cells"><span>3.2 Detection of variable genes across the single cells</span></h3></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px">In [<!-- -->0<!-- -->]:</div><div style="width:100%;border-radius:2px;position:relative;overflow-x:auto;border:1px solid rgb(207, 207, 207)"><pre class="CodeMirror cm-s-default  CodeMirror-wrap" style="width:100%;overflow-x:auto;line-height:normal;height:auto;font-size:inherit;margin-bottom:0;padding:4px;white-space:pre-wrap;word-wrap:break-word;word-break:normal;border:0;background:white"><div style="margin-left:0"><span class="cm-comment">## Here, we select the top 1,000 highly variable genes (Hvg) for downstream analysis.</span><span>
</span><span class="cm-variable">dataset</span><span> </span><span class="cm-operator arrow">&lt;-</span><span> </span><span class="cm-variable">FindVariableFeatures</span><span>(</span><span class="cm-variable">object</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">dataset</span><span>,</span><span> </span><span class="cm-variable">selection.method</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-string">&#x27;</span><span class="cm-string">mean.var.plot&#x27;</span><span>,</span><span> </span><span class="cm-variable">mean.cutoff</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">c</span><span>(</span><span class="cm-number">0.0125</span><span>,</span><span> </span><span class="cm-number">3</span><span>)</span><span>,</span><span> </span><span class="cm-variable">dispersion.cutoff</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">c</span><span>(</span><span class="cm-number">0.5</span><span>,</span><span> </span><span class="cm-atom">Inf</span><span>)</span><span>)</span><span>
</span><span class="cm-variable">length</span><span>(</span><span class="cm-variable">x</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">VariableFeatures</span><span>(</span><span class="cm-variable">object</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">dataset</span><span>)</span><span>)</span><span> </span><span class="cm-comment">#3084</span><span>
</span><span class="cm-comment">## Identify the 10 most highly variable genes</span><span>
</span><span class="cm-variable">top10</span><span> </span><span class="cm-operator arrow">&lt;-</span><span> </span><span class="cm-variable">head</span><span>(</span><span class="cm-variable">VariableFeatures</span><span>(</span><span class="cm-variable">dataset</span><span>)</span><span>,</span><span> </span><span class="cm-number">10</span><span>)</span><span>
</span></div></pre></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px">In [<!-- -->0<!-- -->]:</div><div style="width:100%;border-radius:2px;position:relative;overflow-x:auto;border:1px solid rgb(207, 207, 207)"><pre class="CodeMirror cm-s-default  CodeMirror-wrap" style="width:100%;overflow-x:auto;line-height:normal;height:auto;font-size:inherit;margin-bottom:0;padding:4px;white-space:pre-wrap;word-wrap:break-word;word-break:normal;border:0;background:white"><div style="margin-left:0"><span class="cm-comment">## Plot</span><span>
</span><span class="cm-comment"># plot variable features with and without labels</span><span>
</span><span class="cm-variable">plot1</span><span> </span><span class="cm-operator arrow">&lt;-</span><span> </span><span class="cm-variable">VariableFeaturePlot</span><span>(</span><span class="cm-variable">dataset</span><span>)</span><span>
</span><span class="cm-variable">plot2</span><span> </span><span class="cm-operator arrow">&lt;-</span><span> </span><span class="cm-variable">LabelPoints</span><span>(</span><span class="cm-variable">plot</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">plot1</span><span>,</span><span> </span><span class="cm-variable">points</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">top10</span><span>,</span><span> </span><span class="cm-variable">repel</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-atom">TRUE</span><span>)</span><span>
</span><span class="cm-comment">#plot1</span><span>
</span><span class="cm-variable">plot2</span><span>
</span><span>
</span></div></pre></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px">In [<!-- -->0<!-- -->]:</div><div style="width:100%;border-radius:2px;position:relative;overflow-x:auto;border:1px solid rgb(207, 207, 207)"><pre class="CodeMirror cm-s-default  CodeMirror-wrap" style="width:100%;overflow-x:auto;line-height:normal;height:auto;font-size:inherit;margin-bottom:0;padding:4px;white-space:pre-wrap;word-wrap:break-word;word-break:normal;border:0;background:white"><div style="margin-left:0"><span class="cm-comment"># Generate a vector with the top 1000 hvgens</span><span>
</span><span class="cm-variable">hv.genes</span><span> </span><span class="cm-operator arrow">&lt;-</span><span> </span><span class="cm-variable">head</span><span>(</span><span class="cm-variable">rownames</span><span>(</span><span class="cm-variable">HVFInfo</span><span>(</span><span class="cm-variable">object</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">dataset</span><span>)</span><span>)</span><span>,</span><span> </span><span class="cm-number">1000</span><span>)</span><span>
</span><span>
</span></div></pre></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><h3 id="scale-the-data"><span>Scale the data</span></h3></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><p><span style="text-indent:0"><span>Scale the values for each features. The result is in units of &#x27;standard deviation&#x27; where the mean expression is zero. If a gene is expressed 1 standard deviation more than the mean across all cells, it will have a value of 1. During this process, we can also &#x27;regress out&#x27; technical effects. For instance, the number of features and the mitochondrial percentage.</span></span></p></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px">In [<!-- -->0<!-- -->]:</div><div style="width:100%;border-radius:2px;position:relative;overflow-x:auto;border:1px solid rgb(207, 207, 207)"><pre class="CodeMirror cm-s-default  CodeMirror-wrap" style="width:100%;overflow-x:auto;line-height:normal;height:auto;font-size:inherit;margin-bottom:0;padding:4px;white-space:pre-wrap;word-wrap:break-word;word-break:normal;border:0;background:white"><div style="margin-left:0"><span class="cm-variable">dataset</span><span> </span><span class="cm-operator arrow">&lt;-</span><span> </span><span class="cm-variable">ScaleData</span><span>(</span><span class="cm-variable">object</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">dataset</span><span>,</span><span> </span><span class="cm-variable">features</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">hv.genes</span><span>,</span><span> </span><span class="cm-variable">vars.to.regress</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">c</span><span>(</span><span class="cm-string">&quot;</span><span class="cm-string">nFeature_RNA&quot;</span><span>)</span><span> </span><span>)</span><span> </span><span>
</span></div></pre></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><div><div><div class="alert alert-danger"><h2>Back to the presentation</h2></div></div></div></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><div><div><div class="alert alert-success"><h2>Part IV: Dimensionality reduction techniques</h2>
</div></div></div></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><h3 id="41-perform-dimensionality-reduction"><span>4.1 Perform dimensionality reduction</span></h3></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><p><span style="text-indent:0"><span>You can run PCA, t-SNE and UMAP in simple one-line codes... The parameters are NOT random, but we will ignore this for the sake of time</span></span></p></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px">In [<!-- -->0<!-- -->]:</div><div style="width:100%;border-radius:2px;position:relative;overflow-x:auto;border:1px solid rgb(207, 207, 207)"><pre class="CodeMirror cm-s-default  CodeMirror-wrap" style="width:100%;overflow-x:auto;line-height:normal;height:auto;font-size:inherit;margin-bottom:0;padding:4px;white-space:pre-wrap;word-wrap:break-word;word-break:normal;border:0;background:white"><div style="margin-left:0"><span class="cm-comment"># Dimention reduction</span><span>
</span><span class="cm-variable">dataset</span><span> </span><span class="cm-operator arrow">&lt;-</span><span> </span><span class="cm-variable">RunPCA</span><span>(</span><span class="cm-variable">object</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">dataset</span><span>,</span><span> </span><span class="cm-variable">features</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">hv.genes</span><span>,</span><span> </span><span class="cm-variable">verbose</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-atom">FALSE</span><span>,</span><span class="cm-variable">npcs</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-number">50</span><span>)</span><span>
</span></div></pre></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px">In [<!-- -->0<!-- -->]:</div><div style="width:100%;border-radius:2px;position:relative;overflow-x:auto;border:1px solid rgb(207, 207, 207)"><pre class="CodeMirror cm-s-default  CodeMirror-wrap" style="width:100%;overflow-x:auto;line-height:normal;height:auto;font-size:inherit;margin-bottom:0;padding:4px;white-space:pre-wrap;word-wrap:break-word;word-break:normal;border:0;background:white"><div style="margin-left:0"><span class="cm-variable">dataset</span><span> </span><span class="cm-operator arrow">&lt;-</span><span> </span><span class="cm-variable">RunUMAP</span><span>(</span><span class="cm-variable">object</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">dataset</span><span>,</span><span> </span><span class="cm-variable">reduction</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-string">&quot;</span><span class="cm-string">pca&quot;</span><span>,</span><span> </span><span class="cm-variable">dims</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-number">1</span><span class="cm-operator">:</span><span class="cm-number">50</span><span>,</span><span> </span><span class="cm-variable">verbose</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-atom">FALSE</span><span>)</span><span>
</span><span class="cm-variable">dataset</span><span> </span><span class="cm-operator arrow">&lt;-</span><span> </span><span class="cm-variable">RunTSNE</span><span>(</span><span class="cm-variable">object</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">dataset</span><span>,</span><span> </span><span class="cm-variable">dims</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-number">1</span><span class="cm-operator">:</span><span class="cm-number">50</span><span>,</span><span> </span><span class="cm-variable">verbose</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-atom">FALSE</span><span>)</span><span>
</span></div></pre></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><h3 id="42-examine-and-visualize-pca-results-a-few-different-ways"><span>4.2 Examine and visualize PCA results a few different ways</span></h3></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><h4 id="421-dimplots"><span>4.2.1 Dimplots</span></h4><p><span style="text-indent:0"><span>First, let&#x27;s have a look at the first three dimensions. For this, we will take advantage of some of hte metadata generated by the authors: </span><span><strong>Supercluster</strong></span><span> information</span></span></p></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px">In [<!-- -->0<!-- -->]:</div><div style="width:100%;border-radius:2px;position:relative;overflow-x:auto;border:1px solid rgb(207, 207, 207)"><pre class="CodeMirror cm-s-default  CodeMirror-wrap" style="width:100%;overflow-x:auto;line-height:normal;height:auto;font-size:inherit;margin-bottom:0;padding:4px;white-space:pre-wrap;word-wrap:break-word;word-break:normal;border:0;background:white"><div style="margin-left:0"><span class="cm-variable">options</span><span>(</span><span class="cm-variable">repr.plot.width</span><span class="cm-operator">=</span><span class="cm-number">12</span><span>,</span><span> </span><span class="cm-variable">repr.plot.height</span><span class="cm-operator">=</span><span class="cm-number">6</span><span>)</span><span>
</span><span class="cm-variable">DimPlot</span><span>(</span><span class="cm-variable">dataset</span><span>,</span><span class="cm-variable">cols</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-string">&#x27;</span><span class="cm-string">Spectral&#x27;</span><span>,</span><span class="cm-variable">group.by</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-string">&#x27;</span><span class="cm-string">Supercluster&#x27;</span><span>,</span><span class="cm-variable">pt.size</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-number">1</span><span>,</span><span class="cm-variable">reduction</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-string">&#x27;</span><span class="cm-string">pca&#x27;</span><span>,</span><span class="cm-variable">dims</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">c</span><span>(</span><span class="cm-number">1</span><span>,</span><span class="cm-number">2</span><span>)</span><span>)</span><span>
</span><span class="cm-variable">DimPlot</span><span>(</span><span class="cm-variable">dataset</span><span>,</span><span class="cm-variable">cols</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-string">&#x27;</span><span class="cm-string">Spectral&#x27;</span><span>,</span><span class="cm-variable">group.by</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-string">&#x27;</span><span class="cm-string">Supercluster&#x27;</span><span>,</span><span class="cm-variable">pt.size</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-number">1</span><span>,</span><span class="cm-variable">reduction</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-string">&#x27;</span><span class="cm-string">pca&#x27;</span><span>,</span><span class="cm-variable">dims</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">c</span><span>(</span><span class="cm-number">3</span><span>,</span><span class="cm-number">4</span><span>)</span><span>)</span><span>
</span></div></pre></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><div><div><div class="alert alert-warning"><b>Question 4.1
<p>
</p></b> 
    <p>
        1. What differences do you see between different principle components?  
    </p>
    <p> 
        2. Some genes contribute more to the variation then others... What do these genes and principle components tell us?
    </p>
<p><b>Discuss with your neighbours, in groups of 3, for 2 min. We will go through the groups</b></p>
</div></div></div></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><h4 id="422-list-of-genes-that-influence-pcs-that-most-in-both-+--directions"><span>4.2.2 List of genes that influence PCs that most in both (+/-) directions</span></h4><p><span style="text-indent:0"><span>You see the top genes that contribute to the respective principle components</span></span></p></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px">In [<!-- -->0<!-- -->]:</div><div style="width:100%;border-radius:2px;position:relative;overflow-x:auto;border:1px solid rgb(207, 207, 207)"><pre class="CodeMirror cm-s-default  CodeMirror-wrap" style="width:100%;overflow-x:auto;line-height:normal;height:auto;font-size:inherit;margin-bottom:0;padding:4px;white-space:pre-wrap;word-wrap:break-word;word-break:normal;border:0;background:white"><div style="margin-left:0"><span class="cm-variable">print</span><span>(</span><span class="cm-variable">dataset</span><span>[</span><span>[</span><span class="cm-string">&quot;</span><span class="cm-string">pca&quot;</span><span>]</span><span>]</span><span>,</span><span> </span><span class="cm-variable">dims</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-number">1</span><span class="cm-operator">:</span><span class="cm-number">5</span><span>,</span><span> </span><span class="cm-variable">nfeatures</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-number">5</span><span>)</span><span>
</span></div></pre></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><h4 id="423-visualise-genes-that-influence-pcs-the-most"><span>4.2.3 Visualise genes that influence PCs the most</span></h4></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px">In [<!-- -->0<!-- -->]:</div><div style="width:100%;border-radius:2px;position:relative;overflow-x:auto;border:1px solid rgb(207, 207, 207)"><pre class="CodeMirror cm-s-default  CodeMirror-wrap" style="width:100%;overflow-x:auto;line-height:normal;height:auto;font-size:inherit;margin-bottom:0;padding:4px;white-space:pre-wrap;word-wrap:break-word;word-break:normal;border:0;background:white"><div style="margin-left:0"><span class="cm-variable">VizDimLoadings</span><span>(</span><span class="cm-variable">dataset</span><span>,</span><span> </span><span class="cm-variable">dims</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-number">1</span><span class="cm-operator">:</span><span class="cm-number">2</span><span>,</span><span> </span><span class="cm-variable">reduction</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-string">&quot;</span><span class="cm-string">pca&quot;</span><span>)</span><span>
</span></div></pre></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><h4 id="423-heatmaps-to-visualise-genes-that-influence-pcs"><span>4.2.3 Heatmaps to visualise genes that influence PCs</span></h4></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px">In [<!-- -->0<!-- -->]:</div><div style="width:100%;border-radius:2px;position:relative;overflow-x:auto;border:1px solid rgb(207, 207, 207)"><pre class="CodeMirror cm-s-default  CodeMirror-wrap" style="width:100%;overflow-x:auto;line-height:normal;height:auto;font-size:inherit;margin-bottom:0;padding:4px;white-space:pre-wrap;word-wrap:break-word;word-break:normal;border:0;background:white"><div style="margin-left:0"><span class="cm-variable">options</span><span>(</span><span class="cm-variable">repr.plot.width</span><span class="cm-operator">=</span><span class="cm-number">12</span><span>,</span><span> </span><span class="cm-variable">repr.plot.height</span><span class="cm-operator">=</span><span class="cm-number">9</span><span>)</span><span>
</span><span class="cm-variable">DimHeatmap</span><span>(</span><span class="cm-variable">dataset</span><span>,</span><span> </span><span class="cm-variable">dims</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-number">1</span><span class="cm-operator">:</span><span class="cm-number">4</span><span>,</span><span> </span><span class="cm-variable">cells</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-number">500</span><span>,</span><span> </span><span class="cm-variable">balanced</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-atom">TRUE</span><span>,</span><span> </span><span class="cm-variable">ncol</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-number">2</span><span>)</span><span>
</span></div></pre></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><h3 id="43-determine-statistically-significant-principal-components"><span>4.3 Determine statistically significant principal components</span></h3><p><span style="text-indent:0"><span><strong>How many principle components do we need? There are some practical guidelines that will help with your analysis</strong></span><span><span style="white-space:pre">
</span><span></span></span><span></span></span></p><h4 id="431-jackstrawplot"><span>4.3.1 JackStrawPlot()</span></h4><p><span style="text-indent:0"><span>We can do this with JackStrawPlot() function of Seurat. This is powerful, but time consuming. Running it altogether slows the server considerably. Also, we cannot visualise all different PCs (only up to 25). We will skip it for now (you can run it during the practice if you want to) and instead look at PCs using heatmaps</span></span></p></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%;border-radius:2px;position:relative;overflow-x:auto;border:1px solid rgb(207, 207, 207)"><pre class="CodeMirror cm-s-default  CodeMirror-wrap" style="width:100%;overflow-x:auto;line-height:normal;height:auto;font-size:inherit;margin-bottom:0;padding:4px;white-space:pre-wrap;word-wrap:break-word;word-break:normal;border:0;background:white"><div style="margin-left:0"><span class="cm-variable">dataset</span><span> </span><span class="cm-operator arrow">&lt;-</span><span> </span><span class="cm-variable">JackStraw</span><span>(</span><span class="cm-variable">object</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">dataset</span><span>,</span><span> </span><span class="cm-variable">num.replicate</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-number">100</span><span>)</span><span>
</span><span class="cm-variable">dataset</span><span> </span><span class="cm-operator arrow">&lt;-</span><span> </span><span class="cm-variable">ScoreJackStraw</span><span>(</span><span class="cm-variable">object</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">dataset</span><span>,</span><span> </span><span class="cm-variable">dims</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-number">1</span><span class="cm-operator">:</span><span class="cm-number">20</span><span>)</span><span>
</span><span class="cm-variable">JackStrawPlot</span><span>(</span><span class="cm-variable">object</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">dataset</span><span>,</span><span> </span><span class="cm-variable">dims</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-number">1</span><span class="cm-operator">:</span><span class="cm-number">20</span><span>)</span><span>
</span><span>
</span></div></pre></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><h4 id="432-heatmaps"><span>4.3.2 Heatmaps</span></h4></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><p><span style="text-indent:0"><span>We will plot the top genes from selected PCs and see how they differe between cells. </span></span></p></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px">In [<!-- -->0<!-- -->]:</div><div style="width:100%;border-radius:2px;position:relative;overflow-x:auto;border:1px solid rgb(207, 207, 207)"><pre class="CodeMirror cm-s-default  CodeMirror-wrap" style="width:100%;overflow-x:auto;line-height:normal;height:auto;font-size:inherit;margin-bottom:0;padding:4px;white-space:pre-wrap;word-wrap:break-word;word-break:normal;border:0;background:white"><div style="margin-left:0"><span class="cm-variable">options</span><span>(</span><span class="cm-variable">repr.plot.width</span><span class="cm-operator">=</span><span class="cm-number">18</span><span>,</span><span> </span><span class="cm-variable">repr.plot.height</span><span class="cm-operator">=</span><span class="cm-number">9</span><span>)</span><span>
</span><span class="cm-variable">DimHeatmap</span><span>(</span><span class="cm-variable">dataset</span><span>,</span><span> </span><span class="cm-variable">dims</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">c</span><span>(</span><span class="cm-number">1</span><span>,</span><span class="cm-number">2</span><span>,</span><span class="cm-number">5</span><span>,</span><span class="cm-number">10</span><span>,</span><span>
</span><span>                             </span><span class="cm-number">15</span><span>,</span><span class="cm-number">20</span><span>,</span><span class="cm-number">25</span><span>,</span><span class="cm-number">30</span><span>,</span><span>
</span><span>                             </span><span class="cm-number">35</span><span>,</span><span class="cm-number">40</span><span>,</span><span class="cm-number">45</span><span>,</span><span class="cm-number">50</span><span>)</span><span>,</span><span> </span><span class="cm-variable">cells</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-number">250</span><span>,</span><span> </span><span class="cm-variable">balanced</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-atom">TRUE</span><span>,</span><span> </span><span class="cm-variable">ncol</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-number">4</span><span>)</span><span>
</span></div></pre></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><div><div><div class="alert alert-warning"><b>Question 4.2
<p>What do you see? 
</p></b> 
<p>1. When do you think the variation is meaningful, and when not?  </p>
<p> 2. Some genes contribute more to the variation then others... What do these genes and principle components tell us?
    </p>
<p>3. What does it mean in terms of &#x27;Superclusters&#x27;? </p>
    <p>
         4.How would our results change if we include less or more principle components for downstream analysis?
    </p>
<p><b>Think for yourself for 2 min, then talk to your neighbors, in groups of 3, for 2 min. We will go through the groups</b></p>
</div></div></div></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><h4 id="422-elbowplot-as-an-alternative"><span>4.2.2 Elbowplot as an alternative</span></h4></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px">In [<!-- -->0<!-- -->]:</div><div style="width:100%;border-radius:2px;position:relative;overflow-x:auto;border:1px solid rgb(207, 207, 207)"><pre class="CodeMirror cm-s-default  CodeMirror-wrap" style="width:100%;overflow-x:auto;line-height:normal;height:auto;font-size:inherit;margin-bottom:0;padding:4px;white-space:pre-wrap;word-wrap:break-word;word-break:normal;border:0;background:white"><div style="margin-left:0"><span class="cm-comment"># Plots the standard deviations (or approximate singular values if running PCAFast) of the principle components for easy identification of an elbow in the graph. This elbow often corresponds well with the significant dims and speeds up running than Jackstraw</span><span>
</span><span class="cm-comment"># In this example, it looks like the elbow would fall around PC 11, then 25 and 35 again. After this, there is little change. </span><span>
</span><span class="cm-variable">options</span><span>(</span><span class="cm-variable">repr.plot.width</span><span class="cm-operator">=</span><span class="cm-number">12</span><span>,</span><span> </span><span class="cm-variable">repr.plot.height</span><span class="cm-operator">=</span><span class="cm-number">6</span><span>)</span><span>
</span><span class="cm-variable">ElbowPlot</span><span>(</span><span class="cm-variable">object</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">dataset</span><span>,</span><span> </span><span class="cm-variable">reduction</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-string">&quot;</span><span class="cm-string">pca&quot;</span><span>,</span><span class="cm-variable">ndims</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-number">50</span><span>)</span><span>
</span></div></pre></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><div><div><div class="alert alert-warning"><b>Question 4.3
<p>
</p></b> 
    <p>
        1. What does it mean in terms of &#x27;Superclusters&#x27;? 
    </p>
    <p>
        2. When do you think the variation in a PC is meaningful, and when not?  
    </p>
    <p>
        3.How would our results change if we include less or more principle components for downstream analysis?
    </p>
<p><b>Think for yourself for 2 min, then talk to your neighbors, in groups of 3, for 2 min. We will go through the groups</b></p>
</div></div></div></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><div><div><div class="alert alert-danger"><h2>Back to the presentation</h2></div></div></div></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><div><div><div class="alert alert-success"><h2>Part V: Clustering</h2>
</div></div></div></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><h3 id="51-find-clusters"><span>5.1 Find clusters</span></h3></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><p><span style="text-indent:0"><span>here, we will use graph-based clustering to identify groups of cells that are similar, which we will call &#x27;clusters&#x27;</span></span></p></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><h4 id="511-use-embeded-&#x27;louvain&#x27;-algorithm-for-clustering"><span>5.1.1 Use embeded &#x27;louvain&#x27; algorithm for clustering</span></h4></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px">In [<!-- -->0<!-- -->]:</div><div style="width:100%;border-radius:2px;position:relative;overflow-x:auto;border:1px solid rgb(207, 207, 207)"><pre class="CodeMirror cm-s-default  CodeMirror-wrap" style="width:100%;overflow-x:auto;line-height:normal;height:auto;font-size:inherit;margin-bottom:0;padding:4px;white-space:pre-wrap;word-wrap:break-word;word-break:normal;border:0;background:white"><div style="margin-left:0"><span class="cm-comment">#SNN Graph Construction</span><span>
</span><span class="cm-variable">dataset</span><span> </span><span class="cm-operator arrow">&lt;-</span><span> </span><span class="cm-variable">FindNeighbors</span><span>(</span><span class="cm-variable">object</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">dataset</span><span>,</span><span> </span><span class="cm-variable">dims</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-number">1</span><span class="cm-operator">:</span><span class="cm-number">35</span><span>)</span><span> </span><span class="cm-comment">#cell the pcs based on Jackstraw and Elbow plots</span><span>
</span><span class="cm-comment">#Find clusters using the louvain algorithm</span><span>
</span><span class="cm-variable">dataset</span><span> </span><span class="cm-operator arrow">&lt;-</span><span> </span><span class="cm-variable">FindClusters</span><span>(</span><span class="cm-variable">object</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">dataset</span><span>,</span><span> </span><span class="cm-variable">resolution</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-number">1</span><span>)</span><span> </span><span class="cm-comment"># changing the resolution will change the size/number of clusters! c(0.6, 0.8, 1)</span><span>
</span></div></pre></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><h4 id="512-visualize-qcs"><span>5.1.2 Visualize QCs</span></h4><p><span style="text-indent:0"><span>Let&#x27;s have a look at the QC stats agian. As default, this time the information will be displayed per cluster</span></span></p></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px">In [<!-- -->0<!-- -->]:</div><div style="width:100%;border-radius:2px;position:relative;overflow-x:auto;border:1px solid rgb(207, 207, 207)"><pre class="CodeMirror cm-s-default  CodeMirror-wrap" style="width:100%;overflow-x:auto;line-height:normal;height:auto;font-size:inherit;margin-bottom:0;padding:4px;white-space:pre-wrap;word-wrap:break-word;word-break:normal;border:0;background:white"><div style="margin-left:0"><span class="cm-comment"># Statistics</span><span>
</span><span class="cm-variable">options</span><span>(</span><span class="cm-variable">repr.plot.width</span><span class="cm-operator">=</span><span class="cm-number">12</span><span>,</span><span> </span><span class="cm-variable">repr.plot.height</span><span class="cm-operator">=</span><span class="cm-number">6</span><span>)</span><span>
</span><span class="cm-variable">VlnPlot</span><span>(</span><span class="cm-variable">object</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">dataset</span><span>,</span><span> </span><span class="cm-variable">features</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">c</span><span>(</span><span class="cm-string">&quot;</span><span class="cm-string">nCount_RNA&quot;</span><span>,</span><span> </span><span class="cm-string">&quot;</span><span class="cm-string">nFeature_RNA&quot;</span><span>)</span><span>,</span><span class="cm-variable">pt.size</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-number">0</span><span>)</span><span>
</span></div></pre></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><h3 id="52-cell-type-identification"><span>5.2 Cell type identification</span></h3></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><h4 id="521-visualise-lineage-specific-genes-curated-from-the-literature"><span>5.2.1 Visualise lineage specific genes curated from the literature</span></h4></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><p><span style="text-indent:0"><span>We provide some markers from the literature as a reference for identifying cell types. You can find more of these in Siletti et al 2022 Bioarxiv manuscript</span></span></p><blockquote><p><span style="text-indent:0"><span>ELAVL2, TUBB3:		Neurons </span><span><span style="white-space:pre">
</span><span></span></span><span>PLP1:						Oligodendrocytes </span><span><span style="white-space:pre">
</span><span></span></span><span>GFAP:						Astrocytes </span><span><span style="white-space:pre">
</span><span></span></span><span>ITGAM:						Microglia </span><span><span style="white-space:pre">
</span><span></span></span><span>PECAM1:				Endothelial cells</span></span></p></blockquote></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><h5 id="5211-t-sne-plots-for-visualising-markers-at-single-cell-level"><span>5.2.1.1 t-SNE plots for visualising markers at single cell level</span></h5></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px">In [<!-- -->0<!-- -->]:</div><div style="width:100%;border-radius:2px;position:relative;overflow-x:auto;border:1px solid rgb(207, 207, 207)"><pre class="CodeMirror cm-s-default  CodeMirror-wrap" style="width:100%;overflow-x:auto;line-height:normal;height:auto;font-size:inherit;margin-bottom:0;padding:4px;white-space:pre-wrap;word-wrap:break-word;word-break:normal;border:0;background:white"><div style="margin-left:0"><span class="cm-variable">options</span><span>(</span><span class="cm-variable">repr.plot.width</span><span class="cm-operator">=</span><span class="cm-number">18</span><span>,</span><span> </span><span class="cm-variable">repr.plot.height</span><span class="cm-operator">=</span><span class="cm-number">12</span><span>)</span><span>
</span><span class="cm-variable">ColorRamp</span><span> </span><span class="cm-operator arrow">&lt;-</span><span> </span><span class="cm-variable">colorRampPalette</span><span>(</span><span class="cm-variable">rev</span><span>(</span><span class="cm-variable">brewer.pal</span><span>(</span><span class="cm-variable">n</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-number">7</span><span>,</span><span class="cm-variable">name</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-string">&quot;</span><span class="cm-string">RdYlBu&quot;</span><span>)</span><span>)</span><span>)</span><span>(</span><span class="cm-number">100</span><span>)</span><span> </span><span class="cm-comment"># This is how to generate a color key</span><span>
</span><span class="cm-variable">FeaturePlot</span><span>(</span><span class="cm-variable">object</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">dataset</span><span>,</span><span> </span><span class="cm-variable">features</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">c</span><span>(</span><span class="cm-string">&quot;</span><span class="cm-string">ELAVL2&quot;</span><span>,</span><span class="cm-string">&quot;</span><span class="cm-string">TUBB3&quot;</span><span>,</span><span class="cm-string">&quot;</span><span class="cm-string">PLP1&quot;</span><span>,</span><span class="cm-string">&quot;</span><span class="cm-string">GFAP&quot;</span><span>,</span><span class="cm-string">&quot;</span><span class="cm-string">ITGAM&quot;</span><span>,</span><span class="cm-string">&quot;</span><span class="cm-string">PECAM1&quot;</span><span>)</span><span>,</span><span class="cm-variable">reduction</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-string">&#x27;</span><span class="cm-string">tsne&#x27;</span><span>,</span><span class="cm-variable">cols</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">ColorRamp</span><span>,</span><span> </span><span class="cm-variable">ncol</span><span> </span><span class="cm-operator">=</span><span class="cm-number">3</span><span>)</span><span>
</span></div></pre></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><p><span style="text-indent:0"><span>Neurons are very diverse! We can use additonal markers to classify them functionally as &#x27;Glutamatergic&#x27; and &#x27;GABAergic&#x27;</span></span></p><blockquote><p><span style="text-indent:0"><span>SLC17A5, SLC17A6:				Glutamatergic neurons </span><span><span style="white-space:pre">
</span><span></span></span><span>GAD1, GAD2, SLC32A1:		GABAergic neurons</span></span></p></blockquote></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px">In [<!-- -->0<!-- -->]:</div><div style="width:100%;border-radius:2px;position:relative;overflow-x:auto;border:1px solid rgb(207, 207, 207)"><pre class="CodeMirror cm-s-default  CodeMirror-wrap" style="width:100%;overflow-x:auto;line-height:normal;height:auto;font-size:inherit;margin-bottom:0;padding:4px;white-space:pre-wrap;word-wrap:break-word;word-break:normal;border:0;background:white"><div style="margin-left:0"><span class="cm-variable">options</span><span>(</span><span class="cm-variable">repr.plot.width</span><span class="cm-operator">=</span><span class="cm-number">18</span><span>,</span><span> </span><span class="cm-variable">repr.plot.height</span><span class="cm-operator">=</span><span class="cm-number">12</span><span>)</span><span>
</span><span class="cm-variable">ColorRamp</span><span> </span><span class="cm-operator arrow">&lt;-</span><span> </span><span class="cm-variable">colorRampPalette</span><span>(</span><span class="cm-variable">rev</span><span>(</span><span class="cm-variable">brewer.pal</span><span>(</span><span class="cm-variable">n</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-number">7</span><span>,</span><span class="cm-variable">name</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-string">&quot;</span><span class="cm-string">RdYlBu&quot;</span><span>)</span><span>)</span><span>)</span><span>(</span><span class="cm-number">100</span><span>)</span><span> </span><span class="cm-comment"># This is how to generate a color key</span><span>
</span><span class="cm-variable">FeaturePlot</span><span>(</span><span class="cm-variable">object</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">dataset</span><span>,</span><span> </span><span class="cm-variable">features</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">c</span><span>(</span><span class="cm-string">&quot;</span><span class="cm-string">ELAVL2&quot;</span><span>,</span><span class="cm-string">&quot;</span><span class="cm-string">GAD1&quot;</span><span>,</span><span class="cm-string">&quot;</span><span class="cm-string">GAD2&quot;</span><span>,</span><span class="cm-string">&quot;</span><span class="cm-string">SLC17A5&quot;</span><span>,</span><span class="cm-string">&quot;</span><span class="cm-string">SLC17A6&quot;</span><span>,</span><span class="cm-string">&quot;</span><span class="cm-string">SLC32A1&quot;</span><span>)</span><span>,</span><span class="cm-variable">reduction</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-string">&#x27;</span><span class="cm-string">tsne&#x27;</span><span>,</span><span class="cm-variable">cols</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">ColorRamp</span><span>,</span><span> </span><span class="cm-variable">ncol</span><span> </span><span class="cm-operator">=</span><span class="cm-number">3</span><span>)</span><span>
</span></div></pre></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><h5 id="5212-violin-plot-for-cluster-level-average"><span>5.2.1.2 Violin plot for cluster level average</span></h5></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px">In [<!-- -->0<!-- -->]:</div><div style="width:100%;border-radius:2px;position:relative;overflow-x:auto;border:1px solid rgb(207, 207, 207)"><pre class="CodeMirror cm-s-default  CodeMirror-wrap" style="width:100%;overflow-x:auto;line-height:normal;height:auto;font-size:inherit;margin-bottom:0;padding:4px;white-space:pre-wrap;word-wrap:break-word;word-break:normal;border:0;background:white"><div style="margin-left:0"><span class="cm-variable">options</span><span>(</span><span class="cm-variable">repr.plot.width</span><span class="cm-operator">=</span><span class="cm-number">18</span><span>,</span><span> </span><span class="cm-variable">repr.plot.height</span><span class="cm-operator">=</span><span class="cm-number">6</span><span>)</span><span>
</span><span class="cm-variable">VlnPlot</span><span>(</span><span class="cm-variable">object</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">dataset</span><span>,</span><span> </span><span class="cm-variable">features</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">c</span><span>(</span><span class="cm-string">&quot;</span><span class="cm-string">ELAVL2&quot;</span><span>,</span><span class="cm-string">&quot;</span><span class="cm-string">TUBB3&quot;</span><span>,</span><span class="cm-string">&quot;</span><span class="cm-string">PLP1&quot;</span><span>,</span><span class="cm-string">&quot;</span><span class="cm-string">GFAP&quot;</span><span>,</span><span class="cm-string">&quot;</span><span class="cm-string">ITGAM&quot;</span><span>,</span><span class="cm-string">&quot;</span><span class="cm-string">PECAM1&quot;</span><span>)</span><span>,</span><span class="cm-variable">pt.size</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-number">0</span><span>,</span><span> </span><span class="cm-variable">ncol</span><span> </span><span class="cm-operator">=</span><span class="cm-number">2</span><span>)</span><span> </span><span class="cm-comment">#, slot = &#x27;counts&#x27;</span><span>
</span><span class="cm-comment"># you can plot raw counts as well</span><span>
</span></div></pre></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px">In [<!-- -->0<!-- -->]:</div><div style="width:100%;border-radius:2px;position:relative;overflow-x:auto;border:1px solid rgb(207, 207, 207)"><pre class="CodeMirror cm-s-default  CodeMirror-wrap" style="width:100%;overflow-x:auto;line-height:normal;height:auto;font-size:inherit;margin-bottom:0;padding:4px;white-space:pre-wrap;word-wrap:break-word;word-break:normal;border:0;background:white"><div style="margin-left:0"><span class="cm-variable">options</span><span>(</span><span class="cm-variable">repr.plot.width</span><span class="cm-operator">=</span><span class="cm-number">18</span><span>,</span><span> </span><span class="cm-variable">repr.plot.height</span><span class="cm-operator">=</span><span class="cm-number">6</span><span>)</span><span>
</span><span class="cm-variable">VlnPlot</span><span>(</span><span class="cm-variable">object</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">dataset</span><span>,</span><span> </span><span class="cm-variable">features</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">c</span><span>(</span><span class="cm-string">&quot;</span><span class="cm-string">ELAVL2&quot;</span><span>,</span><span class="cm-string">&quot;</span><span class="cm-string">GAD1&quot;</span><span>,</span><span class="cm-string">&quot;</span><span class="cm-string">GAD2&quot;</span><span>,</span><span class="cm-string">&quot;</span><span class="cm-string">SLC17A5&quot;</span><span>,</span><span class="cm-string">&quot;</span><span class="cm-string">SLC17A6&quot;</span><span>,</span><span class="cm-string">&quot;</span><span class="cm-string">SLC32A1&quot;</span><span>)</span><span>,</span><span class="cm-variable">pt.size</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-number">0</span><span>,</span><span> </span><span class="cm-variable">ncol</span><span> </span><span class="cm-operator">=</span><span class="cm-number">2</span><span>)</span><span> </span><span class="cm-comment">#, slot = &#x27;counts&#x27;</span><span>
</span><span class="cm-comment"># you can plot raw counts as well</span><span>
</span></div></pre></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><p><span style="text-indent:0"><span>You can use marker gene information to name each cluster as </span></span></p><blockquote><p><span style="text-indent:0"><span>cell classes e.g.  </span><span><strong>oligodendrocytes, neurons </strong></span><span> or </span><span><span style="white-space:pre">
</span><span></span></span><span>marker-defined cell types e.g. </span><span><strong>GlutamatergicNeuron_0</strong></span></span></p></blockquote></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><h4 id="522-compare-our-clusters-with-the-metadata"><span>5.2.2 Compare our clusters with the metadata</span></h4><p><span style="text-indent:0"><span>Your clusters are saved in the metadata, as part of the seurat object, with the name &quot;seurat_clusters&quot;. We compare this to the metadata from the authors, which will help you decide which clusters belong to which cell type</span></span></p></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px">In [<!-- -->0<!-- -->]:</div><div style="width:100%;border-radius:2px;position:relative;overflow-x:auto;border:1px solid rgb(207, 207, 207)"><pre class="CodeMirror cm-s-default  CodeMirror-wrap" style="width:100%;overflow-x:auto;line-height:normal;height:auto;font-size:inherit;margin-bottom:0;padding:4px;white-space:pre-wrap;word-wrap:break-word;word-break:normal;border:0;background:white"><div style="margin-left:0"><span class="cm-variable">plot21</span><span> </span><span class="cm-operator arrow">&lt;-</span><span> </span><span class="cm-variable">DimPlot</span><span>(</span><span class="cm-variable">object</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">dataset</span><span>,</span><span> </span><span class="cm-variable">reduction</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-string">&#x27;</span><span class="cm-string">tsne&#x27;</span><span>,</span><span class="cm-variable">group.by</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-string">&quot;</span><span class="cm-string">seurat_clusters&quot;</span><span>)</span><span>
</span><span class="cm-variable">plot22</span><span> </span><span class="cm-operator arrow">&lt;-</span><span> </span><span class="cm-variable">DimPlot</span><span>(</span><span class="cm-variable">object</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">dataset</span><span>,</span><span> </span><span class="cm-variable">reduction</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-string">&#x27;</span><span class="cm-string">tsne&#x27;</span><span>,</span><span class="cm-variable">group.by</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-string">&quot;</span><span class="cm-string">Supercluster&quot;</span><span>)</span><span>
</span><span class="cm-variable">plot23</span><span> </span><span class="cm-operator arrow">&lt;-</span><span> </span><span class="cm-variable">DimPlot</span><span>(</span><span class="cm-variable">object</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">dataset</span><span>,</span><span> </span><span class="cm-variable">reduction</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-string">&#x27;</span><span class="cm-string">tsne&#x27;</span><span>,</span><span class="cm-variable">group.by</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-string">&quot;</span><span class="cm-string">Clusters&quot;</span><span>)</span><span>
</span><span>
</span><span class="cm-comment">#plot, adjusting the size of the graph</span><span>
</span><span class="cm-variable">options</span><span>(</span><span class="cm-variable">repr.plot.width</span><span class="cm-operator">=</span><span class="cm-number">18</span><span>,</span><span> </span><span class="cm-variable">repr.plot.height</span><span class="cm-operator">=</span><span class="cm-number">6</span><span>)</span><span>
</span><span class="cm-variable">plot21</span><span> </span><span class="cm-operator">|</span><span> </span><span class="cm-variable">plot22</span><span>
</span><span class="cm-variable">options</span><span>(</span><span class="cm-variable">repr.plot.width</span><span class="cm-operator">=</span><span class="cm-number">12</span><span>,</span><span> </span><span class="cm-variable">repr.plot.height</span><span class="cm-operator">=</span><span class="cm-number">6</span><span>)</span><span>
</span><span class="cm-variable">plot23</span><span>
</span></div></pre></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><p><span style="text-indent:0"><span>We can visualise the number of cells labeled with each label using the table() function</span></span></p></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px">In [<!-- -->0<!-- -->]:</div><div style="width:100%;border-radius:2px;position:relative;overflow-x:auto;border:1px solid rgb(207, 207, 207)"><pre class="CodeMirror cm-s-default  CodeMirror-wrap" style="width:100%;overflow-x:auto;line-height:normal;height:auto;font-size:inherit;margin-bottom:0;padding:4px;white-space:pre-wrap;word-wrap:break-word;word-break:normal;border:0;background:white"><div style="margin-left:0"><span class="cm-variable">table</span><span>(</span><span class="cm-variable">dataset</span><span>@</span><span class="cm-variable">meta.data</span><span>[</span><span>[</span><span class="cm-string">&#x27;</span><span class="cm-string">Supercluster&#x27;</span><span>]</span><span>]</span><span>)</span><span>
</span></div></pre></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><div><div><div class="alert alert-warning">
    <b>Question 5.1</b>
<p explanation="">
    </p><p questions="">
        1. Why are there such big differences in the number of clusters between different analysis?
    </p>
<p><b how="" to="" answer="">Shoot your answer!</b></p>
</div></div></div></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><h4 id="523-visualise-how-different-labels-compare-to-each-other"><span>5.2.3 Visualise how different labels compare to each other</span></h4></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><p><span style="text-indent:0"><span>To compare them, we will do a trick. First, generate a dataframe with teh information you need, then use the table() function of make a confusion matrix</span></span></p></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px">In [<!-- -->0<!-- -->]:</div><div style="width:100%;border-radius:2px;position:relative;overflow-x:auto;border:1px solid rgb(207, 207, 207)"><pre class="CodeMirror cm-s-default  CodeMirror-wrap" style="width:100%;overflow-x:auto;line-height:normal;height:auto;font-size:inherit;margin-bottom:0;padding:4px;white-space:pre-wrap;word-wrap:break-word;word-break:normal;border:0;background:white"><div style="margin-left:0"><span class="cm-variable">df</span><span> </span><span class="cm-operator arrow">&lt;-</span><span> </span><span class="cm-variable">as.data.frame</span><span>(</span><span class="cm-variable">cbind</span><span>(</span><span class="cm-variable">dataset</span><span>@</span><span class="cm-variable">meta.data</span><span>[</span><span>[</span><span class="cm-string">&#x27;</span><span class="cm-string">Supercluster&#x27;</span><span>]</span><span>]</span><span>,</span><span class="cm-variable">dataset</span><span>@</span><span class="cm-variable">meta.data</span><span>[</span><span>[</span><span class="cm-string">&#x27;</span><span class="cm-string">seurat_clusters&#x27;</span><span>]</span><span>]</span><span>)</span><span>)</span><span>
</span><span class="cm-variable">rownames</span><span>(</span><span class="cm-variable">df</span><span>)</span><span> </span><span class="cm-operator arrow">&lt;-</span><span> </span><span class="cm-variable">colnames</span><span>(</span><span class="cm-variable">dataset</span><span>)</span><span>
</span><span class="cm-variable">colnames</span><span>(</span><span class="cm-variable">df</span><span>)</span><span> </span><span class="cm-operator arrow">&lt;-</span><span> </span><span class="cm-variable">c</span><span>(</span><span class="cm-string">&#x27;</span><span class="cm-string">Supercluster&#x27;</span><span>,</span><span class="cm-string">&#x27;</span><span class="cm-string">seurat_clusters&#x27;</span><span>)</span><span> </span><span class="cm-comment">#Supercluster Clusters</span><span>
</span></div></pre></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px">In [<!-- -->0<!-- -->]:</div><div style="width:100%;border-radius:2px;position:relative;overflow-x:auto;border:1px solid rgb(207, 207, 207)"><pre class="CodeMirror cm-s-default  CodeMirror-wrap" style="width:100%;overflow-x:auto;line-height:normal;height:auto;font-size:inherit;margin-bottom:0;padding:4px;white-space:pre-wrap;word-wrap:break-word;word-break:normal;border:0;background:white"><div style="margin-left:0"><span class="cm-variable">head</span><span>(</span><span class="cm-variable">table</span><span>(</span><span class="cm-variable">df</span><span>)</span><span>[</span><span>,</span><span class="cm-number">1</span><span class="cm-operator">:</span><span class="cm-number">10</span><span>]</span><span>)</span><span> </span><span class="cm-comment">#head() automatically looks at teh first 6 rows. We also shorthen the columns to 10</span><span>
</span></div></pre></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><p><span style="text-indent:0"><span>Quick hetmap to visualise which cell types belong to which cluster</span></span></p></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px">In [<!-- -->0<!-- -->]:</div><div style="width:100%;border-radius:2px;position:relative;overflow-x:auto;border:1px solid rgb(207, 207, 207)"><pre class="CodeMirror cm-s-default  CodeMirror-wrap" style="width:100%;overflow-x:auto;line-height:normal;height:auto;font-size:inherit;margin-bottom:0;padding:4px;white-space:pre-wrap;word-wrap:break-word;word-break:normal;border:0;background:white"><div style="margin-left:0"><span class="cm-variable">heatmap</span><span>(</span><span class="cm-variable">table</span><span>(</span><span class="cm-variable">df</span><span>)</span><span>,</span><span class="cm-variable">xlab</span><span> </span><span class="cm-operator">=</span><span class="cm-string">&#x27;</span><span class="cm-string">seurat_clusters&#x27;</span><span>)</span><span>
</span></div></pre></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><div><div><div class="alert alert-warning">
    <b>Question 5.2</b>
<p explanation="">
    </p><p questions="">
        1. What does the heatmap tell you?
    </p>
<p><b how="" to="" answer="">Shoot your answer!</b></p>
</div></div></div></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><div><div><div class="alert alert-danger"><h2>Back to the presentation</h2></div></div></div></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><div><div><div class="alert alert-success"><h2>Part VI: Differential gene expression analysis</h2>
An important step in scRNAseq data analysis is to find out the molecular signature (or genes) that discriminate a cluster from others. There are different ways of comparing clusters (e.g. two clusters, or one versus all, or a few clusters versus a few clusters) and different statistical methods to achieve this
</div></div></div></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><h3 id="61-find-marker-genes-for-each-cluster"><span>6.1 find marker genes for each cluster</span></h3></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><h4 id="611-start-with-a-single-cluster"><span>6.1.1 Start with a single cluster</span></h4></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><p><span style="text-indent:0"><span>We often want to see what makes a cluster special, or different than the rest. You can achieve this here</span></span></p></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px">In [<!-- -->0<!-- -->]:</div><div style="width:100%;border-radius:2px;position:relative;overflow-x:auto;border:1px solid rgb(207, 207, 207)"><pre class="CodeMirror cm-s-default  CodeMirror-wrap" style="width:100%;overflow-x:auto;line-height:normal;height:auto;font-size:inherit;margin-bottom:0;padding:4px;white-space:pre-wrap;word-wrap:break-word;word-break:normal;border:0;background:white"><div style="margin-left:0"><span class="cm-variable">cluster1.markers</span><span> </span><span class="cm-operator arrow">&lt;-</span><span> </span><span class="cm-variable">FindMarkers</span><span>(</span><span class="cm-variable">object</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">dataset</span><span>,</span><span> </span><span class="cm-variable">ident.1</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-number">0</span><span>,</span><span> </span><span class="cm-variable">min.pct</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-number">0.25</span><span>)</span><span> </span><span class="cm-comment">#, ident.2 = c(0, 3) to compare</span><span>
</span><span class="cm-variable">head</span><span>(</span><span class="cm-variable">x</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">cluster1.markers</span><span>,</span><span> </span><span class="cm-variable">n</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-number">5</span><span>)</span><span>
</span></div></pre></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><h4 id="612-find-markers-for-every-cluster-compared-to-all-remaining-cells-report-only-the-positive-ones"><span>6.1.2 Find markers for every cluster compared to all remaining cells, report only the positive ones</span></h4><p><span style="text-indent:0"><span>This is extremely useful to define what these cells are! However, you will see that a &#x27;global comparison&#x27; is not always as accurate as you would expect</span></span></p></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px">In [<!-- -->0<!-- -->]:</div><div style="width:100%;border-radius:2px;position:relative;overflow-x:auto;border:1px solid rgb(207, 207, 207)"><pre class="CodeMirror cm-s-default  CodeMirror-wrap" style="width:100%;overflow-x:auto;line-height:normal;height:auto;font-size:inherit;margin-bottom:0;padding:4px;white-space:pre-wrap;word-wrap:break-word;word-break:normal;border:0;background:white"><div style="margin-left:0"><span class="cm-variable">library</span><span>(</span><span class="cm-variable">dplyr</span><span>)</span><span>
</span><span class="cm-variable">dataset.markers</span><span> </span><span class="cm-operator arrow">&lt;-</span><span> </span><span class="cm-variable">FindAllMarkers</span><span>(</span><span class="cm-variable">object</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">dataset</span><span>,</span><span> </span><span class="cm-variable">only.pos</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-atom">TRUE</span><span>,</span><span> </span><span class="cm-variable">min.pct</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-number">0.25</span><span>,</span><span> </span><span class="cm-variable">logfc.threshold</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-number">0.25</span><span>)</span><span>
</span><span class="cm-comment"># we have just calculated the marker genes for each cell type</span><span>
</span></div></pre></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><p><span style="text-indent:0"><span>We can select the top genes for each cluster. Here, we will select the top 2 genes</span></span></p></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px">In [<!-- -->0<!-- -->]:</div><div style="width:100%;border-radius:2px;position:relative;overflow-x:auto;border:1px solid rgb(207, 207, 207)"><pre class="CodeMirror cm-s-default  CodeMirror-wrap" style="width:100%;overflow-x:auto;line-height:normal;height:auto;font-size:inherit;margin-bottom:0;padding:4px;white-space:pre-wrap;word-wrap:break-word;word-break:normal;border:0;background:white"><div style="margin-left:0"><span class="cm-variable">dataset.markers</span><span> </span><span class="cm-operator variable-2">%&gt;%</span><span> </span><span class="cm-variable">group_by</span><span>(</span><span class="cm-variable">cluster</span><span>)</span><span> </span><span class="cm-operator variable-2">%&gt;%</span><span> </span><span class="cm-variable">top_n</span><span>(</span><span class="cm-variable">n</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-number">2</span><span>,</span><span> </span><span class="cm-variable">wt</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">avg_log2FC</span><span>)</span><span>  </span><span class="cm-operator variable-2">%&gt;%</span><span> </span><span class="cm-variable">head</span><span>(</span><span class="cm-variable">n</span><span class="cm-operator">=</span><span class="cm-number">10</span><span>)</span><span>
</span><span class="cm-comment"># Takes a bit of time</span><span>
</span></div></pre></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><p><span style="text-indent:0"><span><strong> save the data as separate sheets in an excel file</strong></span></span></p></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px">In [<!-- -->0<!-- -->]:</div><div style="width:100%;border-radius:2px;position:relative;overflow-x:auto;border:1px solid rgb(207, 207, 207)"><pre class="CodeMirror cm-s-default  CodeMirror-wrap" style="width:100%;overflow-x:auto;line-height:normal;height:auto;font-size:inherit;margin-bottom:0;padding:4px;white-space:pre-wrap;word-wrap:break-word;word-break:normal;border:0;background:white"><div style="margin-left:0"><span class="cm-variable">library</span><span>(</span><span class="cm-variable">openxlsx</span><span>)</span><span>
</span><span class="cm-comment">## Create Workbook object and add worksheets</span><span>
</span><span class="cm-variable">wb</span><span> </span><span class="cm-operator arrow">&lt;-</span><span> </span><span class="cm-variable">createWorkbook</span><span>(</span><span>)</span><span>
</span><span class="cm-comment">## Add the cluster info</span><span>
</span><span class="cm-keyword">for</span><span> </span><span>(</span><span> </span><span class="cm-variable">n</span><span> </span><span class="cm-keyword">in</span><span> </span><span class="cm-variable">unique</span><span>(</span><span class="cm-variable">dataset.markers</span><span class="cm-operator dollar">$</span><span class="cm-variable">cluster</span><span>)</span><span> </span><span>)</span><span> </span><span>{</span><span>
</span><span>  </span><span class="cm-variable">addWorksheet</span><span>(</span><span class="cm-variable">wb</span><span>,</span><span> </span><span class="cm-variable">n</span><span>)</span><span>
</span><span>  </span><span class="cm-variable">writeData</span><span>(</span><span class="cm-variable">wb</span><span>,</span><span> </span><span class="cm-variable">n</span><span>,</span><span> </span><span class="cm-variable">dataset.markers</span><span>[</span><span class="cm-variable">dataset.markers</span><span class="cm-operator dollar">$</span><span class="cm-variable">cluster</span><span> </span><span class="cm-operator">==</span><span> </span><span class="cm-variable">n</span><span>,</span><span>]</span><span>,</span><span> </span><span class="cm-variable">startCol</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-number">1</span><span>,</span><span> </span><span class="cm-variable">startRow</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-number">1</span><span>,</span><span> </span><span class="cm-variable">rowNames</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-atom">TRUE</span><span>,</span><span> </span><span class="cm-variable">colNames</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-atom">TRUE</span><span>)</span><span>
</span><span>}</span><span>
</span><span class="cm-comment">#save</span><span>
</span><span class="cm-variable">saveWorkbook</span><span>(</span><span class="cm-variable">wb</span><span>,</span><span> </span><span class="cm-variable">paste</span><span>(</span><span class="cm-variable">name</span><span>,</span><span class="cm-string">&quot;</span><span class="cm-string">_cell_clust_diff_genes.xlsx&quot;</span><span>,</span><span class="cm-variable">sep</span><span class="cm-operator">=</span><span class="cm-string">&quot;</span><span class="cm-string">&quot;</span><span>)</span><span>,</span><span> </span><span class="cm-variable">overwrite</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-atom">TRUE</span><span>)</span><span>
</span></div></pre></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><h3 id="62-doheatmap-generates-an-expression-heatmap-for-given-cells-and-features-"><span>6.2 DoHeatmap generates an expression heatmap for given cells and features. </span></h3><p><span style="text-indent:0"><span>In this case, we are plotting the top 20 markers (or all markers if less than 20) for each cluster.</span></span></p></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px">In [<!-- -->0<!-- -->]:</div><div style="width:100%;border-radius:2px;position:relative;overflow-x:auto;border:1px solid rgb(207, 207, 207)"><pre class="CodeMirror cm-s-default  CodeMirror-wrap" style="width:100%;overflow-x:auto;line-height:normal;height:auto;font-size:inherit;margin-bottom:0;padding:4px;white-space:pre-wrap;word-wrap:break-word;word-break:normal;border:0;background:white"><div style="margin-left:0"><span class="cm-variable">dataset.markers</span><span> </span><span class="cm-operator variable-2">%&gt;%</span><span> </span><span class="cm-variable">group_by</span><span>(</span><span class="cm-variable">cluster</span><span>)</span><span> </span><span class="cm-operator variable-2">%&gt;%</span><span> </span><span class="cm-variable">top_n</span><span>(</span><span class="cm-variable">n</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-number">10</span><span>,</span><span> </span><span class="cm-variable">wt</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">avg_log2FC</span><span>)</span><span> </span><span class="cm-operator arrow">-&gt;</span><span> </span><span class="cm-variable">top10</span><span>
</span></div></pre></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px">In [<!-- -->0<!-- -->]:</div><div style="width:100%;border-radius:2px;position:relative;overflow-x:auto;border:1px solid rgb(207, 207, 207)"><pre class="CodeMirror cm-s-default  CodeMirror-wrap" style="width:100%;overflow-x:auto;line-height:normal;height:auto;font-size:inherit;margin-bottom:0;padding:4px;white-space:pre-wrap;word-wrap:break-word;word-break:normal;border:0;background:white"><div style="margin-left:0"><span class="cm-variable">options</span><span>(</span><span class="cm-variable">repr.plot.width</span><span class="cm-operator">=</span><span class="cm-number">12</span><span>,</span><span> </span><span class="cm-variable">repr.plot.height</span><span class="cm-operator">=</span><span class="cm-number">12</span><span>)</span><span>
</span><span class="cm-comment">#dataset &lt;- ScaleData(object = dataset, features = rownames(top10$gene)) #,vars.to.regress = &quot;percent.mt&quot;</span><span>
</span><span class="cm-variable">DoHeatmap</span><span>(</span><span class="cm-variable">object</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">dataset</span><span>,</span><span> </span><span class="cm-variable">features</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">top10</span><span class="cm-operator dollar">$</span><span class="cm-variable">gene</span><span>[</span><span class="cm-number">1</span><span class="cm-operator">:</span><span class="cm-number">100</span><span>]</span><span>,</span><span class="cm-variable">slot</span><span> </span><span class="cm-operator">=</span><span>  </span><span class="cm-string">&#x27;</span><span class="cm-string">scale.data&#x27;</span><span>)</span><span> </span><span class="cm-operator">+</span><span> </span><span class="cm-variable">NoLegend</span><span>(</span><span>)</span><span>
</span></div></pre></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><div><div><div class="alert alert-warning"><b>Question 1
<p>
What do you see? :</p></b> 
<p>1. How do the number of genes and RNA molecules correlate?  </p>
<p>2. There are two populations visible on the violin plot... What do these mean? </p>
<p><b>Answer whenever you want!</b></p>
</div></div></div></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><div><div><div class="alert alert-danger"><h2>Back to the presentation</h2></div></div></div></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><div><div><div class="alert alert-success"><h2>Assignment</h2>
 <h5>
     Workflow
    </h5>
1. Discuss a workplan, split the work     <br/>
2. Go through the assignment individually <br/>
3. Come together and discuss your part with each others <br/>
4. Prepare a very short presentation <br/>
<br/>
You can use Monday afternoon, as well as the practice session on Tuesday morning. The instructor will be available for questions through Slack on Monday, and in person at the practice session on Tuesday morning. We will have the time to wrap-up the summary at the end of the practice<br/><br/>
 <h5>
     To Do
    </h5>
This section is different for each group
<br/><br/></div></div></div><p><span style="text-indent:0"><span>More information:</span><span><span style="white-space:normal"> </span><span></span></span><a href="https://satijalab.org/seurat/" target="_blank" rel="noopener"><span>https://satijalab.org/seurat/</span></a><span></span></span></p><p><span style="text-indent:0"><span></span><a href="https://satijalab.org/seurat/v3.1/visualization_vignette.html" target="_blank" rel="noopener"><span>https://satijalab.org/seurat/v3.1/visualization_vignette.html</span></a><span></span></span></p></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><h2 id="extra-code-for-the-practical"><span>Extra code for the practical</span></h2></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><h4 id="subsetting-neuronal-clusters-&#x27;0&#x27;-and-&#x27;1&#x27;"><span>subsetting neuronal clusters &#x27;0&#x27; and &#x27;1&#x27;</span></h4></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px">In [<!-- -->0<!-- -->]:</div><div style="width:100%;border-radius:2px;position:relative;overflow-x:auto;border:1px solid rgb(207, 207, 207)"><pre class="CodeMirror cm-s-default  CodeMirror-wrap" style="width:100%;overflow-x:auto;line-height:normal;height:auto;font-size:inherit;margin-bottom:0;padding:4px;white-space:pre-wrap;word-wrap:break-word;word-break:normal;border:0;background:white"><div style="margin-left:0"><span class="cm-variable">small</span><span> </span><span class="cm-operator arrow">&lt;-</span><span> </span><span class="cm-variable">subset</span><span>(</span><span class="cm-variable">dataset</span><span>,</span><span> </span><span class="cm-variable">idents</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-variable">c</span><span>(</span><span class="cm-string">&#x27;</span><span class="cm-string">0&#x27;</span><span>,</span><span class="cm-string">&#x27;</span><span class="cm-string">1&#x27;</span><span>)</span><span>)</span><span>
</span></div></pre></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px">In [<!-- -->0<!-- -->]:</div><div style="width:100%;border-radius:2px;position:relative;overflow-x:auto;border:1px solid rgb(207, 207, 207)"><pre class="CodeMirror cm-s-default  CodeMirror-wrap" style="width:100%;overflow-x:auto;line-height:normal;height:auto;font-size:inherit;margin-bottom:0;padding:4px;white-space:pre-wrap;word-wrap:break-word;word-break:normal;border:0;background:white"><div style="margin-left:0"><span class="cm-variable">small</span><span>
</span></div></pre></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px">In [<!-- -->0<!-- -->]:</div><div style="width:100%;border-radius:2px;position:relative;overflow-x:auto;border:1px solid rgb(207, 207, 207)"><pre class="CodeMirror cm-s-default  CodeMirror-wrap" style="width:100%;overflow-x:auto;line-height:normal;height:auto;font-size:inherit;margin-bottom:0;padding:4px;white-space:pre-wrap;word-wrap:break-word;word-break:normal;border:0;background:white"><div style="margin-left:0"><span class="cm-variable">DimPlot</span><span>(</span><span class="cm-variable">small</span><span>,</span><span> </span><span class="cm-variable">reduction</span><span> </span><span class="cm-operator">=</span><span> </span><span class="cm-string">&#x27;</span><span class="cm-string">tsne&#x27;</span><span>)</span><span>
</span></div></pre></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><p><span style="text-indent:0"><span>You can proceed from step 3.2 to repeat the analysis of a subset of hte dataset that contains clusters 1 and 2</span></span></p></div></div><div></div></div><div style="margin-bottom:10px"><div style="display:flex;flex-direction:row;align-items:stretch"><div style="color:#303F9F;min-width:7em;font-family:monospace;text-align:right;padding-right:5px"></div><div style="width:100%"><h1 id="the-end"><span>THE END</span></h1></div></div><div></div></div></div></div>
</body>
